drivers/timer/nrf_grtc_timer.c

```diff
@@ -0,0 +1,569 @@
+{
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		return 0;
+	}
+
+	return (uint32_t)(counter_sub(counter(), last_count) / CYC_PER_TICK);
+}
+
+static int sys_clock_driver_init(void)
+{
+	nrfx_err_t err_code;
+
+#if defined(CONFIG_NRF_GRTC_TIMER_CLOCK_MANAGEMENT) &&                                             \
+	(defined(NRF_GRTC_HAS_CLKSEL) && (NRF_GRTC_HAS_CLKSEL == 1))
+	/* Use System LFCLK as the low-frequency clock source. */
+	nrfy_grtc_clksel_set(NRF_GRTC, NRF_GRTC_CLKSEL_LFCLK);
+#endif
+
+#if defined(CONFIG_NRF_GRTC_START_SYSCOUNTER)
+	/* SYSCOUNTER needs to be turned off before initialization. */
+	nrfy_grtc_sys_counter_set(NRF_GRTC, false);
+	nrfy_grtc_timeout_set(NRF_GRTC, TIMEOUT);
+	nrfy_grtc_waketime_set(NRF_GRTC, WAKETIME);
+#endif /* CONFIG_NRF_GRTC_START_SYSCOUNTER */
+
+	IRQ_CONNECT(DT_IRQN(GRTC_NODE), DT_IRQ(GRTC_NODE, priority), nrfx_grtc_irq_handler, 0, 0);
+
+	err_code = nrfx_grtc_init(0);
+	if (err_code != NRFX_SUCCESS) {
+		return -EPERM;
+	}
+
+#if defined(CONFIG_NRF_GRTC_START_SYSCOUNTER)
+	err_code = nrfx_grtc_syscounter_start(true, &system_clock_channel_data.channel);
+	if (err_code != NRFX_SUCCESS) {
+		return err_code == NRFX_ERROR_NO_MEM ? -ENOMEM : -EPERM;
+	}
+	if (IS_ENABLED(CONFIG_NRF_GRTC_SLEEP_ALLOWED)) {
+		nrfy_grtc_sys_counter_auto_mode_set(NRF_GRTC, false);
+	}
+#else
+	err_code = nrfx_grtc_channel_alloc(&system_clock_channel_data.channel);
+	if (err_code != NRFX_SUCCESS) {
+		return -ENOMEM;
+	}
+#endif /* CONFIG_NRF_GRTC_START_SYSCOUNTER */
+
+	if (!IS_ENABLED(CONFIG_NRF_GRTC_SLEEP_ALLOWED)) {
+		grtc_active_set();
+	}
+
+	int_mask = NRFX_GRTC_CONFIG_ALLOWED_CC_CHANNELS_MASK;
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		system_timeout_set(CYC_PER_TICK);
+	}
+
+#if CONFIG_SYSTEM_CLOCK_WAIT
+	static const enum nrf_lfclk_start_mode mode =
+		IS_ENABLED(CONFIG_SYSTEM_CLOCK_NO_WAIT)
+			? CLOCK_CONTROL_NRF_LF_START_NOWAIT
+			: (IS_ENABLED(CONFIG_SYSTEM_CLOCK_WAIT_FOR_AVAILABILITY)
+				   ? CLOCK_CONTROL_NRF_LF_START_AVAILABLE
+				   : CLOCK_CONTROL_NRF_LF_START_STABLE);
+
+	z_nrf_clock_control_lf_on(mode);
+#endif
+
+	return 0;
+}
+
+void sys_clock_set_timeout(int32_t ticks, bool idle)
+{
+	ARG_UNUSED(idle);
+	uint64_t cyc, off, now;
+
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		return;
+	}
+
+	ticks = (ticks == K_TICKS_FOREVER) ? MAX_TICKS : MIN(MAX_TICKS, MAX(ticks - 1, 0));
+
+	now = counter();
+
+	/* Round up to the next tick boundary */
+	off = (now - last_count) + (CYC_PER_TICK - 1);
+	off = (off / CYC_PER_TICK) * CYC_PER_TICK;
+
+	/* Get the offset with respect to now */
+	off -= (now - last_count);
+
+	/* Add the offset to get to the next tick boundary */
+	cyc = (uint64_t)ticks * CYC_PER_TICK + off;
+
+	/* Due to elapsed time the calculation above might produce a
+	 * duration that laps the counter.  Don't let it.
+	 */
+	if (cyc > MAX_CYCLES) {
+		cyc = MAX_CYCLES;
+	}
+
+	system_timeout_set(cyc == 0 ? 1 : cyc);
+}
+
+#if defined(CONFIG_NRF_GRTC_TIMER_APP_DEFINED_INIT)
```

Comment:
Old driver supports `sys_clock_disable` which is useful for lowering power consumption and for booting other images, is that not possible here?

Reply:
It will be looked into, but not in scope of this PR.